`SessionConnectorTest.testConcurrentAccess()` fails intermittently on JDK8 and Ubuntu

I've seen this fail a few times but IIRC only on the JDK8 ubuntu-latest runner. Other JDKs or OS seem to be unaffected.
There is a logging exception occurring right before the error.

```
2024-08-22T03:57:24.3602352Z [INFO] Running quickfix.mina.SessionConnectorTest
2024-08-22T03:57:24.4732033Z java.lang.ArrayIndexOutOfBoundsException: 366
2024-08-22T03:57:24.4774887Z 	at java.util.ArrayList.add(ArrayList.java:465)
2024-08-22T03:57:24.4776022Z 	at quickfix.SLF4JLogTest$TestHandler.publish(SLF4JLogTest.java:243)
2024-08-22T03:57:24.4777361Z 	at java.util.logging.Logger.log(Logger.java:738)
2024-08-22T03:57:24.4890414Z 	at org.slf4j.jul.JDK14LoggerAdapter.innerNormalizedLoggingCallHandler(JDK14LoggerAdapter.java:156)
2024-08-22T03:57:24.4892073Z 	at org.slf4j.jul.JDK14LoggerAdapter.log(JDK14LoggerAdapter.java:172)
2024-08-22T03:57:24.4893113Z 	at quickfix.SLF4JLog.log(SLF4JLog.java:140)
2024-08-22T03:57:24.4893904Z 	at quickfix.SLF4JLog.onEvent(SLF4JLog.java:115)
2024-08-22T03:57:24.4894692Z 	at quickfix.Session.<init>(Session.java:600)
2024-08-22T03:57:24.4895756Z 	at quickfix.DefaultSessionFactory.create(DefaultSessionFactory.java:241)
2024-08-22T03:57:24.4897164Z 	at quickfix.mina.SessionConnector.createSession(SessionConnector.java:182)
2024-08-22T03:57:24.4916824Z 	at quickfix.mina.SessionConnectorTest.lambda$testConcurrentAccess$1(SessionConnectorTest.java:308)
2024-08-22T03:57:24.4919208Z 	at java.lang.Thread.run(Thread.java:750)
2024-08-22T03:57:24.5479068Z ##[error]Exception in thread "Client_143" java.lang.AssertionError: This operation shouldn't fail
2024-08-22T03:57:24.5490053Z 	at org.junit.Assert.fail(Assert.java:89)
2024-08-22T03:57:24.5491157Z 	at quickfix.mina.SessionConnectorTest.lambda$testConcurrentAccess$1(SessionConnectorTest.java:315)
2024-08-22T03:57:24.5492311Z 	at java.lang.Thread.run(Thread.java:750)
2024-08-22T03:57:24.6975991Z [ERROR] Tests run: 5, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.336 s <<< FAILURE! -- in quickfix.mina.SessionConnectorTest
2024-08-22T03:57:24.6977426Z [ERROR] quickfix.mina.SessionConnectorTest.testConcurrentAccess -- Time elapsed: 0.266 s <<< FAILURE!
2024-08-22T03:57:24.6978100Z java.lang.AssertionError: 
2024-08-22T03:57:24.6978837Z We should have all sessions exported. Failure here means initialisation has failed somewhere
2024-08-22T03:57:24.6979876Z Expected: <1002>
2024-08-22T03:57:24.6980242Z      but: was <1000>
2024-08-22T03:57:24.6980641Z 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
2024-08-22T03:57:24.6981364Z 	at org.junit.Assert.assertThat(Assert.java:964)
2024-08-22T03:57:24.6982512Z 	at quickfix.mina.SessionConnectorTest.testConcurrentAccess(SessionConnectorTest.java:330)
2024-08-22T03:57:24.6983737Z 	at java.lang.reflect.Method.invoke(Method.java:498)
```

--- a/quickfixj-core/src/test/java/quickfix/SLF4JLogTest.java
+++ b/quickfixj-core/src/test/java/quickfix/SLF4JLogTest.java
@@ -61,18 +61,22 @@ public class SLF4JLogTest {
         setUpLoggerForTest(SLF4JLog.DEFAULT_EVENT_CATEGORY);
         log.onEvent(loggedText);
         assertMessageLogged(SLF4JLog.DEFAULT_EVENT_CATEGORY, sessionID, loggedText);
+        removeLogHandlers(SLF4JLog.DEFAULT_EVENT_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY);
         log.onErrorEvent(loggedText);
         assertMessageLogged(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY, sessionID, loggedText);
-
+        removeLogHandlers(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY);
+        
         setUpLoggerForTest(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
         log.onIncoming(loggedText);
         assertMessageLogged(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY, sessionID, loggedText);
+        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
         log.onOutgoing(loggedText);
         assertMessageLogged(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY, sessionID, loggedText);
+        removeLogHandlers(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
 
         settings.setString(sessionID, SLF4JLogFactory.SETTING_EVENT_CATEGORY, "event");
         settings.setString(sessionID, SLF4JLogFactory.SETTING_ERROR_EVENT_CATEGORY, "errorEvent");
@@ -83,18 +87,22 @@ public class SLF4JLogTest {
         setUpLoggerForTest("event");
         log.onEvent(loggedText);
         assertMessageLogged("event", sessionID, loggedText);
+        removeLogHandlers("event");
 
         setUpLoggerForTest("errorEvent");
         log.onErrorEvent(loggedText);
         assertMessageLogged("errorEvent", sessionID, loggedText);
+        removeLogHandlers("errorEvent");
 
         setUpLoggerForTest("in");
         log.onIncoming(loggedText);
         assertMessageLogged("in", sessionID, loggedText);
+        removeLogHandlers("in");
 
         setUpLoggerForTest("out");
         log.onOutgoing(loggedText);
         assertMessageLogged("out", sessionID, loggedText);
+        removeLogHandlers("out");
     }
 
     @Test
@@ -112,10 +120,12 @@ public class SLF4JLogTest {
         setUpLoggerForTest(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
         log.onIncoming(loggedText);
         assertMessageLogged(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY, sessionID, loggedText);
+        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
         log.onOutgoing(loggedText);
         assertMessageLogged(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY, sessionID, loggedText);
+        removeLogHandlers(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
 
         settings.setBool(sessionID, SLF4JLogFactory.SETTING_LOG_HEARTBEATS, false);
         log = (SLF4JLog) factory.create(sessionID);
@@ -123,10 +133,12 @@ public class SLF4JLogTest {
         setUpLoggerForTest(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
         log.onIncoming(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
         log.onOutgoing(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
     }
 
     @Test
@@ -145,21 +157,25 @@ public class SLF4JLogTest {
         getTestHandler(SLF4JLog.DEFAULT_EVENT_CATEGORY).setLevel(Level.WARNING);
         log.onEvent(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_EVENT_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_EVENT_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY);
         getTestHandler(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY).setLevel(Level.SEVERE);
         log.onErrorEvent(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_EVENT_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_EVENT_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
         getTestHandler(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY).setLevel(Level.WARNING);
         log.onIncoming(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
         getTestHandler(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY).setLevel(Level.WARNING);
         log.onOutgoing(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
     }
 
     /**
@@ -215,17 +231,25 @@ public class SLF4JLogTest {
 
     private TestHandler setUpLoggerForTest(String category) {
         final Logger logger = Logger.getLogger(category);
-        logger.setUseParentHandlers(false);
-        final Handler[] handlers = logger.getHandlers();
-        for (final Handler handler : handlers) {
-            //System.err.println("Removing unexpected handler: " + handlers[i]);
-            logger.removeHandler(handler);
-        }
+        removeLogHandlers(logger);
         final TestHandler testHandler = new TestHandler();
         logger.addHandler(testHandler);
         return testHandler;
     }
 
+    private void removeLogHandlers(String category) {
+        final Logger logger = Logger.getLogger(category);
+        removeLogHandlers(logger);
+    }
+    
+    private void removeLogHandlers(Logger logger) {
+        logger.setUseParentHandlers(false);
+        final Handler[] handlers = logger.getHandlers();
+        for (final Handler handler : handlers) {
+            logger.removeHandler(handler);
+        }
+    }
+    
     private class TestHandler extends java.util.logging.Handler {
         public final ArrayList<LogRecord> records = new ArrayList<>();
 

--- a/quickfixj-core/src/test/java/quickfix/SLF4JLogTest.java
+++ b/quickfixj-core/src/test/java/quickfix/SLF4JLogTest.java
@@ -44,6 +44,14 @@ public class SLF4JLogTest {
     @After
     public void tearDown() throws Exception {
         SystemTime.setTimeSource(null);
+        removeLogHandlers(SLF4JLog.DEFAULT_EVENT_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
+        removeLogHandlers("event");
+        removeLogHandlers("errorEvent");
+        removeLogHandlers("in");
+        removeLogHandlers("out");
     }
 
     @Test
@@ -95,16 +103,6 @@ public class SLF4JLogTest {
         setUpLoggerForTest("out");
         log.onOutgoing(loggedText);
         assertMessageLogged("out", sessionID, loggedText);
-
-        // cleanup
-        removeLogHandlers(SLF4JLog.DEFAULT_EVENT_CATEGORY);
-        removeLogHandlers(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY);
-        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
-        removeLogHandlers(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
-        removeLogHandlers("event");
-        removeLogHandlers("errorEvent");
-        removeLogHandlers("in");
-        removeLogHandlers("out");
     }
 
     @Test
@@ -137,10 +135,6 @@ public class SLF4JLogTest {
         setUpLoggerForTest(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
         log.onOutgoing(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
-
-        // cleanup
-        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
-        removeLogHandlers(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
     }
 
     @Test
@@ -174,12 +168,6 @@ public class SLF4JLogTest {
         getTestHandler(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY).setLevel(Level.WARNING);
         log.onOutgoing(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
-
-        // cleanup
-        removeLogHandlers(SLF4JLog.DEFAULT_EVENT_CATEGORY);
-        removeLogHandlers(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY);
-        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
-        removeLogHandlers(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
     }
 
     /**

--- a/quickfixj-core/src/test/java/quickfix/SLF4JLogTest.java
+++ b/quickfixj-core/src/test/java/quickfix/SLF4JLogTest.java
@@ -163,7 +163,7 @@ public class SLF4JLogTest {
         getTestHandler(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY).setLevel(Level.SEVERE);
         log.onErrorEvent(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_EVENT_CATEGORY);
-        removeLogHandlers(SLF4JLog.DEFAULT_EVENT_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
         getTestHandler(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY).setLevel(Level.WARNING);

--- a/quickfixj-core/src/test/java/quickfix/SLF4JLogTest.java
+++ b/quickfixj-core/src/test/java/quickfix/SLF4JLogTest.java
@@ -61,22 +61,18 @@ public class SLF4JLogTest {
         setUpLoggerForTest(SLF4JLog.DEFAULT_EVENT_CATEGORY);
         log.onEvent(loggedText);
         assertMessageLogged(SLF4JLog.DEFAULT_EVENT_CATEGORY, sessionID, loggedText);
-        removeLogHandlers(SLF4JLog.DEFAULT_EVENT_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY);
         log.onErrorEvent(loggedText);
         assertMessageLogged(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY, sessionID, loggedText);
-        removeLogHandlers(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY);
         
         setUpLoggerForTest(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
         log.onIncoming(loggedText);
         assertMessageLogged(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY, sessionID, loggedText);
-        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
         log.onOutgoing(loggedText);
         assertMessageLogged(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY, sessionID, loggedText);
-        removeLogHandlers(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
 
         settings.setString(sessionID, SLF4JLogFactory.SETTING_EVENT_CATEGORY, "event");
         settings.setString(sessionID, SLF4JLogFactory.SETTING_ERROR_EVENT_CATEGORY, "errorEvent");
@@ -87,21 +83,27 @@ public class SLF4JLogTest {
         setUpLoggerForTest("event");
         log.onEvent(loggedText);
         assertMessageLogged("event", sessionID, loggedText);
-        removeLogHandlers("event");
 
         setUpLoggerForTest("errorEvent");
         log.onErrorEvent(loggedText);
         assertMessageLogged("errorEvent", sessionID, loggedText);
-        removeLogHandlers("errorEvent");
 
         setUpLoggerForTest("in");
         log.onIncoming(loggedText);
         assertMessageLogged("in", sessionID, loggedText);
-        removeLogHandlers("in");
 
         setUpLoggerForTest("out");
         log.onOutgoing(loggedText);
         assertMessageLogged("out", sessionID, loggedText);
+
+        // cleanup
+        removeLogHandlers(SLF4JLog.DEFAULT_EVENT_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
+        removeLogHandlers("event");
+        removeLogHandlers("errorEvent");
+        removeLogHandlers("in");
         removeLogHandlers("out");
     }
 
@@ -120,12 +122,10 @@ public class SLF4JLogTest {
         setUpLoggerForTest(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
         log.onIncoming(loggedText);
         assertMessageLogged(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY, sessionID, loggedText);
-        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
         log.onOutgoing(loggedText);
         assertMessageLogged(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY, sessionID, loggedText);
-        removeLogHandlers(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
 
         settings.setBool(sessionID, SLF4JLogFactory.SETTING_LOG_HEARTBEATS, false);
         log = (SLF4JLog) factory.create(sessionID);
@@ -133,11 +133,13 @@ public class SLF4JLogTest {
         setUpLoggerForTest(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
         log.onIncoming(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
-        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
         log.onOutgoing(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
+
+        // cleanup
+        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
         removeLogHandlers(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
     }
 
@@ -157,24 +159,26 @@ public class SLF4JLogTest {
         getTestHandler(SLF4JLog.DEFAULT_EVENT_CATEGORY).setLevel(Level.WARNING);
         log.onEvent(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_EVENT_CATEGORY);
-        removeLogHandlers(SLF4JLog.DEFAULT_EVENT_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY);
         getTestHandler(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY).setLevel(Level.SEVERE);
         log.onErrorEvent(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_EVENT_CATEGORY);
-        removeLogHandlers(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
         getTestHandler(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY).setLevel(Level.WARNING);
         log.onIncoming(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
-        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
 
         setUpLoggerForTest(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
         getTestHandler(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY).setLevel(Level.WARNING);
         log.onOutgoing(loggedText);
         assertMessageNotLogged(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
+
+        // cleanup
+        removeLogHandlers(SLF4JLog.DEFAULT_EVENT_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_ERROR_EVENT_CATEGORY);
+        removeLogHandlers(SLF4JLog.DEFAULT_INCOMING_MSG_CATEGORY);
         removeLogHandlers(SLF4JLog.DEFAULT_OUTGOING_MSG_CATEGORY);
     }
 

